File: aiplayer.cpp
/***************************************************************************
             aiplayer.cpp  -  computer player class for othello
                             -------------------
    begin                : Mon May 5 2003
    copyright            : (C) 2003 by Ilmari Heikkinen
    email                : irheikki@cc.hut.fi
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include <qtimer.h>

#include "aiplayer.h"
#include "gamerules.h"
#include "ai.h"

AIPlayer::AIPlayer(unsigned int color) : QObject(),
rules(new GameRules()), trules(new TRules()), gTimer(new QTimer()), color(color)
{
    opposite = (color==BLACK) ? WHITE : BLACK;
    connect( gTimer, SIGNAL(timeout()), this, SLOT(processMove()) );
}

AIPlayer::~AIPlayer() {
    gTimer->stop();
    
    delete rules;
    delete trules;
    delete gTimer;
}

void AIPlayer::extMove(unsigned int pos) {
      last_pos = pos;
      gTimer->start( 600, TRUE );
}

void AIPlayer::processMove() {
    unsigned int move;
    TCoord tmove;
    if (rules->getTurn() == opposite) {
        if (rules->setPiece(last_pos,opposite))
            trules->rulesSetStone(rules->uiToTCoord(last_pos),opposite);
    }
    if (rules->getTurn() == color) {
      tmove = findMove(*trules, color);
      move = rules->TCoordToUi(tmove);
      if (rules->setPiece(move, color)) {
        trules->rulesSetStone(tmove, color);
        emit makeMove( move );
      }
    }
    if (rules->getTurn() == BLANK) {
      emit disconnectMe( this );
    }
}

File: html/src/app/aiplayer.ts
import { BLANK, BLACK, WHITE } from './constants';
import { GameRules } from './gamerules';
import { TRules } from './rules';
import { findMove } from './ai';

export class AIPlayer {
  private rules = new GameRules();
  private trules = new TRules();
  private color: number;
  private opposite: number;
  private lastPos = 0;
  private timeoutId: number | null = null;

  onMakeMove: ((pos: number) => void) | null = null;
  onDisconnect: (() => void) | null = null;

  constructor(color: number) {
    this.color = color;
    this.opposite = color === BLACK ? WHITE : BLACK;
  }

  extMove(pos: number): void {
    this.lastPos = pos;
    if (this.timeoutId !== null) {
      window.clearTimeout(this.timeoutId);
    }
    this.timeoutId = window.setTimeout(() => this.processMove(), 600);
  }

  private processMove(): void {
    if (this.rules.getTurn() === this.opposite) {
      if (this.rules.setPiece(this.lastPos, this.opposite)) {
        this.trules.rulesSetStone(this.rules.uiToTCoord(this.lastPos), this.opposite);
      }
    }
    if (this.rules.getTurn() === this.color) {
      const tmove = findMove(this.trules, this.color);
      const move = this.rules.TCoordToUi(tmove);
      if (this.rules.setPiece(move, this.color)) {
        this.trules.rulesSetStone(tmove, this.color);
        this.onMakeMove?.(move);
      }
    }
    if (this.rules.getTurn() === BLANK) {
      this.onDisconnect?.();
    }
  }
}
